### ISSUE TITLE
`smooth_samples()` errors when models have both a 1D and 2D smooth

### ISSUE THREAD

**Issue by Aariq** [gavinsimpson/gratia#126]
When `smooth_samples()` is run on a model with both 1 and 2 dimensional smooths (e.g. s() and te() or s(f), and s(x, z)) it errors (relatively uninformatively).

``` r
library(mgcv)
#> Loading required package: nlme
#> This is mgcv 1.8-36. For overview type 'help("mgcv-package")'.
library(gratia)

dat <- data_sim("eg2", n = 1000, dist = "normal", scale = 1, seed = 2)

# fit GAM
m <- gam(y ~ s(f) + s(x, z), data = dat, method = "REML")

# get samples
#works
smooth_samples(m, "s(f)", n = 6) 
#> # A tibble: 1,200 × 7
#>    smooth term  by_variable     .x1   row  draw  value
#>    <chr>  <chr> <chr>         <dbl> <int> <int>  <dbl>
#>  1 s(f)   s(f)  <NA>        0.00435     1     1 -0.335
#>  2 s(f)   s(f)  <NA>        0.00776     2     1 -0.331
#>  3 s(f)   s(f)  <NA>        0.0112      3     1 -0.327
#>  4 s(f)   s(f)  <NA>        0.0146      4     1 -0.323
#>  5 s(f)   s(f)  <NA>        0.0180      5     1 -0.319
#>  6 s(f)   s(f)  <NA>        0.0214      6     1 -0.315
#>  7 s(f)   s(f)  <NA>        0.0248      7     1 -0.311
#>  8 s(f)   s(f)  <NA>        0.0282      8     1 -0.307
#>  9 s(f)   s(f)  <NA>        0.0316      9     1 -0.303
#> 10 s(f)   s(f)  <NA>        0.0350     10     1 -0.299
#> # … with 1,190 more rows
smooth_samples(m, "s(x,z)", n = 6)
#> # A tibble: 1,200 × 7
#>    smooth term   by_variable     .x1   row  draw   value
#>    <chr>  <chr>  <chr>         <dbl> <int> <int>   <dbl>
#>  1 s(x,z) s(x,z) <NA>        0.00435     1     1 -0.0226
#>  2 s(x,z) s(x,z) <NA>        0.00776     2     1 -0.0222
#>  3 s(x,z) s(x,z) <NA>        0.0112      3     1 -0.0219
#>  4 s(x,z) s(x,z) <NA>        0.0146      4     1 -0.0215
#>  5 s(x,z) s(x,z) <NA>        0.0180      5     1 -0.0212
#>  6 s(x,z) s(x,z) <NA>        0.0214      6     1 -0.0209
#>  7 s(x,z) s(x,z) <NA>        0.0248      7     1 -0.0205
#>  8 s(x,z) s(x,z) <NA>        0.0282      8     1 -0.0202
#>  9 s(x,z) s(x,z) <NA>        0.0316      9     1 -0.0199
#> 10 s(x,z) s(x,z) <NA>        0.0350     10     1 -0.0196
#> # … with 1,190 more rows

#fails
smooth_samples(m, n = 6)
#> Error: New columns must be compatible with `.data`.
#> x New column has 80000 rows.
#> ℹ `.data` has 40200 rows.
```

<sup>Created on 2022-01-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.2 (2020-06-22)
#>  os       macOS  10.16                
#>  system   x86_64, darwin17.0          
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/New_York            
#>  date     2022-01-10                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  ! package     * version  date       lib source                              
#>  P assertthat    0.2.1    2019-03-21 [?] CRAN (R 4.0.2)                      
#>  P backports     1.2.1    2020-12-09 [?] CRAN (R 4.0.2)                      
#>  P cli           3.0.1    2021-07-17 [?] CRAN (R 4.0.2)                      
#>  P colorspace    2.0-2    2021-06-24 [?] CRAN (R 4.0.2)                      
#>  P crayon        1.4.1    2021-02-08 [?] CRAN (R 4.0.2)                      
#>  P DBI           1.1.1    2021-01-15 [?] CRAN (R 4.0.2)                      
#>  P digest        0.6.27   2020-10-24 [?] CRAN (R 4.0.2)                      
#>  P dplyr         1.0.7    2021-06-18 [?] CRAN (R 4.0.2)                      
#>  P ellipsis      0.3.2    2021-04-29 [?] CRAN (R 4.0.2)                      
#>  P evaluate      0.14     2019-05-28 [?] CRAN (R 4.0.1)                      
#>  P fansi         0.5.0    2021-05-25 [?] CRAN (R 4.0.2)                      
#>  P fastmap       1.1.0    2021-01-25 [?] CRAN (R 4.0.2)                      
#>  P fs            1.5.0    2020-07-31 [?] CRAN (R 4.0.2)                      
#>  P generics      0.1.0    2020-10-31 [?] CRAN (R 4.0.2)                      
#>  P ggplot2       3.3.5    2021-06-25 [?] CRAN (R 4.0.2)                      
#>  P glue          1.4.2    2020-08-27 [?] CRAN (R 4.0.2)                      
#>  P gratia      * 0.6.9400 2022-01-10 [?] Github (gavinsimpson/gratia@416f12d)
#>  P gtable        0.3.0    2019-03-25 [?] CRAN (R 4.0.2)                      
#>  P highr         0.9      2021-04-16 [?] CRAN (R 4.0.2)                      
#>  P htmltools     0.5.2    2021-08-25 [?] CRAN (R 4.0.2)                      
#>  P knitr         1.36     2021-09-29 [?] CRAN (R 4.0.2)                      
#>  P lattice       0.20-44  2021-05-02 [3] CRAN (R 4.0.2)                      
#>  P lifecycle     1.0.0    2021-02-15 [?] CRAN (R 4.0.2)                      
#>  P magrittr      2.0.1    2020-11-17 [?] CRAN (R 4.0.2)                      
#>  P Matrix        1.3-4    2021-06-01 [3] CRAN (R 4.0.2)                      
#>  P mgcv        * 1.8-36   2021-06-01 [3] CRAN (R 4.0.2)                      
#>  P munsell       0.5.0    2018-06-12 [?] CRAN (R 4.0.2)                      
#>  P mvnfast       0.2.7    2021-05-20 [?] CRAN (R 4.0.2)                      
#>  P nlme        * 3.1-152  2021-02-04 [3] CRAN (R 4.0.2)                      
#>  P patchwork     1.1.1    2020-12-17 [?] CRAN (R 4.0.2)                      
#>  P pillar        1.6.2    2021-07-29 [?] CRAN (R 4.0.2)                      
#>  P pkgconfig     2.0.3    2019-09-22 [?] CRAN (R 4.0.2)                      
#>  P purrr         0.3.4    2020-04-17 [?] CRAN (R 4.0.2)                      
#>  P R6            2.5.1    2021-08-19 [?] CRAN (R 4.0.2)                      
#>  P Rcpp          1.0.7    2021-07-07 [?] CRAN (R 4.0.2)                      
#>  P reprex        2.0.1    2021-08-05 [?] CRAN (R 4.0.2)                      
#>  P rlang         0.4.11   2021-04-30 [?] CRAN (R 4.0.2)                      
#>  P rmarkdown     2.10     2021-08-06 [?] CRAN (R 4.0.2)                      
#>  P rstudioapi    0.13     2020-11-12 [?] CRAN (R 4.0.2)                      
#>  P scales        1.1.1    2020-05-11 [?] CRAN (R 4.0.2)                      
#>    sessioninfo   1.1.1    2018-11-05 [3] CRAN (R 4.0.0)                      
#>  P stringi       1.7.4    2021-08-25 [?] CRAN (R 4.0.2)                      
#>  P stringr       1.4.0    2019-02-10 [?] CRAN (R 4.0.2)                      
#>    styler        1.5.1    2021-07-13 [3] CRAN (R 4.0.2)                      
#>  P tibble        3.1.4    2021-08-25 [?] CRAN (R 4.0.2)                      
#>  P tidyr         1.1.3    2021-03-03 [?] CRAN (R 4.0.2)                      
#>  P tidyselect    1.1.1    2021-04-30 [?] CRAN (R 4.0.2)                      
#>  P utf8          1.2.2    2021-07-24 [?] CRAN (R 4.0.2)                      
#>  P vctrs         0.3.8    2021-04-29 [?] CRAN (R 4.0.2)                      
#>  P withr         2.4.2    2021-04-18 [?] CRAN (R 4.0.2)                      
#>  P xfun          0.25     2021-08-06 [?] CRAN (R 4.0.2)                      
#>  P yaml          2.2.1    2020-02-01 [?] CRAN (R 4.0.2)                      
#> 
#> [1] /Users/scottericr/Documents/heliconia-ipm/renv/library/R-4.0/x86_64-apple-darwin17.0
#> [2] /private/var/folders/b_/2vfnxxls5vs401tmhhb3wqdh0000gp/T/RtmpHrcXIv/renv-system-library
#> [3] /Library/Frameworks/R.framework/Versions/4.0/Resources/library
#> 
#>  P ── Loaded and on-disk path mismatch.
```

</details>

**Comment by Aariq** [comment]
Ah, sorry if this is a duplicate of #121

**Comment by Aariq** [comment]
Another (probably related) bug:  The output of `smooth_samples()` is missing data for a 2D smooth only when the model also has a 1D smooth.

``` r
library(mgcv)
#> Loading required package: nlme
#> This is mgcv 1.8-36. For overview type 'help("mgcv-package")'.
library(gratia)

dat <- data_sim("eg2", n = 1000, dist = "normal", scale = 1, seed = 2)

# Fit GAM
m <- gam(y ~ s(x,z), data = dat)
m2 <- gam(y ~ s(f) + s(x,z), data = dat)

#has predicted values for x and z dimensions
samples1 <- smooth_samples(m, "s(x,z)", n = 5)
".x2" %in% colnames(samples1)
#> [1] TRUE

# missing a dimension
samples2 <- smooth_samples(m2, "s(x,z)", n = 5)
".x2" %in% colnames(samples2)
#> [1] FALSE
```

<sup>Created on 2022-01-10 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

<details style="margin-bottom:10px;">
<summary>
Session info
</summary>

``` r
sessioninfo::session_info()
#> ─ Session info ───────────────────────────────────────────────────────────────
#>  setting  value                       
#>  version  R version 4.0.2 (2020-06-22)
#>  os       macOS  10.16                
#>  system   x86_64, darwin17.0          
#>  ui       X11                         
#>  language (EN)                        
#>  collate  en_US.UTF-8                 
#>  ctype    en_US.UTF-8                 
#>  tz       America/New_York            
#>  date     2022-01-10                  
#> 
#> ─ Packages ───────────────────────────────────────────────────────────────────
#>  ! package     * version  date       lib source                              
#>  P assertthat    0.2.1    2019-03-21 [?] CRAN (R 4.0.2)                      
#>  P backports     1.2.1    2020-12-09 [?] CRAN (R 4.0.2)                      
#>  P cli           3.0.1    2021-07-17 [?] CRAN (R 4.0.2)                      
#>  P colorspace    2.0-2    2021-06-24 [?] CRAN (R 4.0.2)                      
#>  P crayon        1.4.1    2021-02-08 [?] CRAN (R 4.0.2)                      
#>  P DBI           1.1.1    2021-01-15 [?] CRAN (R 4.0.2)                      
#>  P digest        0.6.27   2020-10-24 [?] CRAN (R 4.0.2)                      
#>  P dplyr         1.0.7    2021-06-18 [?] CRAN (R 4.0.2)                      
#>  P ellipsis      0.3.2    2021-04-29 [?] CRAN (R 4.0.2)                      
#>  P evaluate      0.14     2019-05-28 [?] CRAN (R 4.0.1)                      
#>  P fansi         0.5.0    2021-05-25 [?] CRAN (R 4.0.2)                      
#>  P fastmap       1.1.0    2021-01-25 [?] CRAN (R 4.0.2)                      
#>  P fs            1.5.0    2020-07-31 [?] CRAN (R 4.0.2)                      
#>  P generics      0.1.0    2020-10-31 [?] CRAN (R 4.0.2)                      
#>  P ggplot2       3.3.5    2021-06-25 [?] CRAN (R 4.0.2)                      
#>  P glue          1.4.2    2020-08-27 [?] CRAN (R 4.0.2)                      
#>  P gratia      * 0.6.9400 2022-01-10 [?] Github (gavinsimpson/gratia@416f12d)
#>  P gtable        0.3.0    2019-03-25 [?] CRAN (R 4.0.2)                      
#>  P highr         0.9      2021-04-16 [?] CRAN (R 4.0.2)                      
#>  P htmltools     0.5.2    2021-08-25 [?] CRAN (R 4.0.2)                      
#>  P knitr         1.36     2021-09-29 [?] CRAN (R 4.0.2)                      
#>  P lattice       0.20-44  2021-05-02 [3] CRAN (R 4.0.2)                      
#>  P lifecycle     1.0.0    2021-02-15 [?] CRAN (R 4.0.2)                      
#>  P magrittr      2.0.1    2020-11-17 [?] CRAN (R 4.0.2)                      
#>  P Matrix        1.3-4    2021-06-01 [3] CRAN (R 4.0.2)                      
#>  P mgcv        * 1.8-36   2021-06-01 [3] CRAN (R 4.0.2)                      
#>  P munsell       0.5.0    2018-06-12 [?] CRAN (R 4.0.2)                      
#>  P mvnfast       0.2.7    2021-05-20 [?] CRAN (R 4.0.2)                      
#>  P nlme        * 3.1-152  2021-02-04 [3] CRAN (R 4.0.2)                      
#>  P patchwork     1.1.1    2020-12-17 [?] CRAN (R 4.0.2)                      
#>  P pillar        1.6.2    2021-07-29 [?] CRAN (R 4.0.2)                      
#>  P pkgconfig     2.0.3    2019-09-22 [?] CRAN (R 4.0.2)                      
#>  P purrr         0.3.4    2020-04-17 [?] CRAN (R 4.0.2)                      
#>  P R6            2.5.1    2021-08-19 [?] CRAN (R 4.0.2)                      
#>  P Rcpp          1.0.7    2021-07-07 [?] CRAN (R 4.0.2)                      
#>  P reprex        2.0.1    2021-08-05 [?] CRAN (R 4.0.2)                      
#>  P rlang         0.4.11   2021-04-30 [?] CRAN (R 4.0.2)                      
#>  P rmarkdown     2.10     2021-08-06 [?] CRAN (R 4.0.2)                      
#>  P rstudioapi    0.13     2020-11-12 [?] CRAN (R 4.0.2)                      
#>  P scales        1.1.1    2020-05-11 [?] CRAN (R 4.0.2)                      
#>    sessioninfo   1.1.1    2018-11-05 [3] CRAN (R 4.0.0)                      
#>  P stringi       1.7.4    2021-08-25 [?] CRAN (R 4.0.2)                      
#>  P stringr       1.4.0    2019-02-10 [?] CRAN (R 4.0.2)                      
#>    styler        1.5.1    2021-07-13 [3] CRAN (R 4.0.2)                      
#>  P tibble        3.1.4    2021-08-25 [?] CRAN (R 4.0.2)                      
#>  P tidyr         1.1.3    2021-03-03 [?] CRAN (R 4.0.2)                      
#>  P tidyselect    1.1.1    2021-04-30 [?] CRAN (R 4.0.2)                      
#>  P utf8          1.2.2    2021-07-24 [?] CRAN (R 4.0.2)                      
#>  P vctrs         0.3.8    2021-04-29 [?] CRAN (R 4.0.2)                      
#>  P withr         2.4.2    2021-04-18 [?] CRAN (R 4.0.2)                      
#>  P xfun          0.25     2021-08-06 [?] CRAN (R 4.0.2)                      
#>  P yaml          2.2.1    2020-02-01 [?] CRAN (R 4.0.2)                      
#> 
#> [1] /Users/scottericr/Documents/heliconia-ipm/renv/library/R-4.0/x86_64-apple-darwin17.0
#> [2] /private/var/folders/b_/2vfnxxls5vs401tmhhb3wqdh0000gp/T/RtmpHrcXIv/renv-system-library
#> [3] /Library/Frameworks/R.framework/Versions/4.0/Resources/library
#> 
#>  P ── Loaded and on-disk path mismatch.
```

</details>

**Comment by Aariq** [comment]
This is where the error is, FYI
https://github.com/gavinsimpson/gratia/blob/fabcb553aea6813ce8ac77c1ec522a4dd6e2f0e0/R/posterior-samples.R#L420

`nr_simu` is 200L when each smooth run separately, but when run with `term=NULL` it's 200 for the first run through the for loop, then 40,000 for i=2

**Comment by gavinsimpson** [comment]
Thanks @Aariq &mdash; yes, it fails here because I didn't anticipate (in the code) that you'd have smooths of different dimension. The solution is to do as `smooth_estimates()` and `compare_smooths()` do and to treat each smooth separately, and nest the data columns as a data frame. Then we can combine all the samples for the different smooths into a single object, and only then unnest it. I just need to implement this...

**Comment by gavinsimpson** [comment]
Thanks again @Aariq for reporting this. This is fixed in 0.8.0.92 which will hopefully go to CRAN later today. Note though that in following how `smooth_estimates()` and `compare_smooths()` work, `smooth_samples()` no longer returns variables named `.x1`, `.x2` etc. The data variables now take their proper names as per the data frame used when the model was fitted.
